### PR TITLE
IsoBands Part-2: Add support in Contour Layer.

### PR DESCRIPTION
**Pull request by 1chandu**
For #2180 
NOTE: stacked on top of #2429 

contour-lnglat

contour-isobands-lnglat

#### Background
- Add support in ContourLayer for rendering Isobands using SolidPolygon layer. (CPU)
- Update contour utility methods to generate polygons for Isobands.
- Update API and update documentation. (non breaking API change)
- Add rendering tests and add new examples layer-browser.

#### Change List
- IsoBands Part-2: Add support in Contour Layer

### CODE REVIEW

File: modules/layers/src/contour-layer/contour-layer.js [hunk 1]
```diff
@@ -24,6 +24,7 @@ import {
   _pointToDensityGridData as pointToDensityGridData
 } from '@deck.gl/core';
 import {LineLayer} from '@deck.gl/layers';
+import {_SolidPolygonLayer as SolidPolygonLayer} from '@deck.gl/layers';
```
Comment (Pessimistress) [inline modules/layers/src/contour-layer/contour-layer.js]: For this line and above: you are importing from the same module, just use relative path??

File: modules/layers/src/contour-layer/contour-layer.js [hunk 2]
```diff
@@ -35,11 +36,13 @@ const defaultProps = {
   // grid aggregation
   cellSize: {type: 'number', min: 1, max: 1000, value: 1000},
   getPosition: {type: 'accessor', value: x => x.position},
+  getWeight: {type: 'accessor', value: x => 1},
 
   // contour lines
   contours: [{threshold: DEFAULT_THRESHOLD}],
 
-  fp64: false
+  fp64: false,
```
Comment (Pessimistress) [inline modules/layers/src/contour-layer/contour-layer.js]: Just curious, does the aggregator support fp64?

File: modules/layers/src/contour-layer/contour-layer.js [hunk 3]
```diff
@@ -167,11 +165,38 @@ export default class ContourLayer extends CompositeLayer {
     return aggregationFlags;
   }
 
-  onGetSublayerColor(segment) {
+  getLineLayerProps() {
```
Comment (Pessimistress) [inline modules/layers/src/contour-layer/contour-layer.js]: Private method names should start with `_`

File: modules/layers/src/contour-layer/contour-layer.js [hunk 4]
```diff
@@ -192,15 +217,31 @@ export default class ContourLayer extends CompositeLayer {
     return strokeWidth;
   }
 
-  rebuildContours({oldProps, props}) {
-    if (oldProps.contours.length !== props.contours.length) {
+  shouldRebuildContours({oldProps, props}) {
+    if (
+      !oldProps.contours ||
+      !oldProps.zOffsetScale ||
+      oldProps.contours.length !== props.contours.length ||
```
Comment (Pessimistress) [inline modules/layers/src/contour-layer/contour-layer.js]: Is length comparison enough?

File: modules/layers/src/contour-layer/contour-layer.js (hunk 2 above)
Comment (1chandu) [reply modules/layers/src/contour-layer/contour-layer.js]: Yeah, it supports.

File: docs/layers/contour-layer.md [hunk 5]
```diff
@@ -65,11 +67,22 @@ NOTE: GPU Aggregation requires WebGL2 support by the browser. When `gpuAggregati
 
 Array of objects with following keys
 
-* `threshold` (Number) : Threshold value to be used in contour generation.
+* `threshold` (Number or Array) :
+
+  - Isolines: `threshold` value must be a single `Number`, Isolines are generated based on this threshold value.
+  - Isobands: `threshold` value must be an Array of two `Number`s. Isobands are generated using `[threshold[0], threshold[1])` as threshold range. NOTE: `threshold[0]` is inclusive and `threshold[1]` is not inclusive.
+
+* `color` (Array, optional) : RGB color array to be used to render the contour, if not specified a default value of `[255, 255, 255]` is used.
```
Comment (gnavvy) [inline docs/layers/contour-layer.md]: same as above, I think we support alpha channel as well.

File: modules/core/src/experimental/utils/gpu-grid-aggregation/grid-aggregation-utils.js [hunk 6]
```diff
@@ -15,6 +15,7 @@ export function pointToDensityGridData({
   gpuGridAggregator,
   gpuAggregation,
   aggregationFlags,
+  getWeight = null,
```
Comment (gnavvy) [inline modules/core/src/experimental/utils/gpu-grid-aggregation/grid-aggregation-utils.js]: Can we use a defined weight over `null` here? like the default accessors in the layers.
e.g.:
```
getWeight = d => [1.0, 0, 0],
```

File: modules/layers/src/contour-layer/marching-squares-codes.js [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+    0: [NW_TRAPEZOID, SE_TRAPEZOID],
+    1: [OCTAGON],
+    2: [SW_TRAPEZOID, NE_TRAPEZOID]
+  },
+  [ternaryToIndex('0202')]: {
+    0: [NE_TRAPEZOID, SW_TRAPEZOID],
+    1: [OCTAGON],
+    2: [NW_TRAPEZOID, SE_TRAPEZOID]
+  }
+};
```
Comment (gnavvy) [inline modules/layers/src/contour-layer/marching-squares-codes.js]: 👏

File: modules/layers/src/contour-layer/contour-layer.js (hunk 4 above)
Comment (1chandu) [reply modules/layers/src/contour-layer/contour-layer.js]: In the code below, I am also checking for individual contour thresholds.

Also added triggers to detect cases when a color or strokeWidth is changed for a contour. Added unit tests

**Comment by 1chandu** [comment]
> Not in this PR: maybe add a bench test for performance?

Yeah sure, once GPU version is ready.

**Pull request merged**